Clang PCH breaks build with boost

https://github.com/riscv-software-src/riscv-isa-sim/commit/13ce3cb1eedb337eee3f45ffdf2e32332a22d50d breaks build with Apple clang when `boost` installation is detected by `configure`. 
Building on 
```
g++ -MMD -MP -I/opt/homebrew/opt/boost/include -DPREFIX=\"/usr/local\" -Wall -Wno-unused -g -O2 -fPIC -std=c++11 -g -O2   -I. -I.. -I../fesvr -I../riscv -I../disasm -I../customext -I../fdt -I../softfloat -I../spike_main -I../spike_dasm   -include-pch  insn_template.h.gch  -c ../riscv/clint.cc
In file included from ../riscv/interactive.cc:3:
In file included from ../riscv/sim.h:11:
In file included from /opt/homebrew/opt/boost/include/boost/asio.hpp:21:
In file included from /opt/homebrew/opt/boost/include/boost/asio/associated_executor.hpp:22:
In file included from /opt/homebrew/opt/boost/include/boost/asio/system_executor.hpp:20:
In file included from /opt/homebrew/opt/boost/include/boost/asio/execution.hpp:19:
In file included from /opt/homebrew/opt/boost/include/boost/asio/execution/any_executor.hpp:31:
In file included from /opt/homebrew/opt/boost/include/boost/asio/execution/blocking.hpp:25:
In file included from /opt/homebrew/opt/boost/include/boost/asio/prefer.hpp:23:
/opt/homebrew/opt/boost/include/boost/asio/traits/require_free.hpp:60:36: error: too many arguments provided to function-like macro invocation
    decltype(require(declval<T>(), declval<Property>()))
                                   ^
/Volumes/src_cs/riscv/riscv-isa-sim/build/../riscv/decode.h:244:9: note: macro 'require' defined here
#define require(x) do { if (unlikely(!(x))) throw trap_illegal_instruction(insn.bits()); } while (0)
        ^
g++ -MMD -MP -I/opt/homebrew/opt/boost/include -DPREFIX=\"/usr/local\" -Wall -Wno-unused -g -O2 -fPIC -std=c++11 -g -O2   -I. -I.. -I../fesvr -I../riscv -I../disasm -I../customext -I../fdt -I../softfloat -I../spike_main -I../spike_dasm   -include-pch  insn_template.h.gch  -c ../riscv/jtag_dtm.cc
In file included from ../riscv/debug_module.cc:3:
In file included from ../riscv/sim.h:9:
In file included from /opt/homebrew/opt/boost/include/boost/algorithm/string.hpp:19:
In file included from /opt/homebrew/opt/boost/include/boost/algorithm/string/trim.hpp:19:
In file included from /opt/homebrew/opt/boost/include/boost/range/as_literal.hpp:18:
In file included from /opt/homebrew/opt/boost/include/boost/range/iterator_range.hpp:13:
In file included from /opt/homebrew/opt/boost/include/boost/range/iterator_range_core.hpp:27:
In file included from /opt/homebrew/opt/boost/include/boost/iterator/iterator_facade.hpp:13:
In file included from /opt/homebrew/opt/boost/include/boost/iterator/iterator_categories.hpp:16:
In file included from /opt/homebrew/opt/boost/include/boost/mpl/placeholders.hpp:24:
In file included from /opt/homebrew/opt/boost/include/boost/mpl/arg.hpp:25:
In file included from /opt/homebrew/opt/boost/include/boost/mpl/aux_/na_assert.hpp:23:
/opt/homebrew/opt/boost/include/boost/mpl/assert.hpp:174:20: error: expected a qualified name after 'typename'
template< typename P > struct assert_arg_pred
                   ^
/Volumes/src_cs/riscv/riscv-isa-sim/build/../riscv/decode.h:177:11: note: expanded from macro 'P'
#define P (*p)
          ^
```
The problem occurs only if `boost` is picked up _and_ `-include-pch  insn_template.h.gch` flag is present.
Removing `HAVE_CLANG_PCH` and `autoconf` fixes the issue.
I could not detect a drastic increase in the build time either.

```
diff --git a/configure.ac b/configure.ac
index 13797a0..46b19f4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -93,7 +93,8 @@ AC_CHECK_TYPE([__int128_t], AC_SUBST([HAVE_INT128],[yes]))
 
 AX_APPEND_LINK_FLAGS([-Wl,--export-dynamic])
 
-AX_CHECK_COMPILE_FLAG([-relocatable-pch], AC_SUBST([HAVE_CLANG_PCH],[yes]))
+AC_SUBST([HAVE_CLANG_PCH],[])
```
My setup:
macOS Big Sur M1
clang-1300.0.29.3
hombrew boost 1.76.0

--- a/Makefile.in
+++ b/Makefile.in
@@ -38,7 +38,7 @@ scripts_dir  := $(src_dir)/scripts
 
 HAVE_INT128 := @HAVE_INT128@
 HAVE_DLOPEN := @HAVE_DLOPEN@
-HAVE_CLANG_PCH := $(and @HAVE_CLANG_PCH@, $(if @BOOST_ASIO_LIB@,,yes))
+HAVE_CLANG_PCH := @HAVE_CLANG_PCH@
 
 # If the version information is not in the configure script, then we
 # assume that we are in a working directory. We use the vcs-version.sh

--- a/customext/cflush.cc
+++ b/customext/cflush.cc
@@ -1,3 +1,4 @@
+#include "insn_macros.h"
 #include "extension.h"
 #include <cstring>
 

--- a/riscv/csrs.cc
+++ b/riscv/csrs.cc
@@ -8,6 +8,8 @@
 #include "decode.h"
 // For trap_virtual_instruction and trap_illegal_instruction:
 #include "trap.h"
+// For require():
+#include "insn_macros.h"
 
 // STATE macro used by require_privilege() macro:
 #undef STATE

--- a/riscv/decode.h
+++ b/riscv/decode.h
@@ -174,7 +174,6 @@ private:
 // helpful macros, etc
 #define MMU (*p->get_mmu())
 #define STATE (*p->get_state())
-#define P (*p)
 #define FLEN (p->get_flen())
 #define READ_REG(reg) STATE.XPR[reg]
 #define READ_FREG(reg) STATE.FPR[reg]
@@ -241,7 +240,6 @@ private:
 #define get_field(reg, mask) (((reg) & (decltype(reg))(mask)) / ((mask) & ~((mask) << 1)))
 #define set_field(reg, mask, val) (((reg) & ~(decltype(reg))(mask)) | (((decltype(reg))(val) * ((mask) & ~((mask) << 1))) & (decltype(reg))(mask)))
 
-#define require(x) do { if (unlikely(!(x))) throw trap_illegal_instruction(insn.bits()); } while (0)
 #define require_privilege(p) require(STATE.prv >= (p))
 #define require_novirt() if (unlikely(STATE.v)) throw trap_virtual_instruction(insn.bits())
 #define require_rv64 require(xlen == 64)

--- a/riscv/insn_macros.h
+++ b/riscv/insn_macros.h
@@ -1,0 +1,9 @@
+#ifndef _RISCV_INSN_MACROS_H
+#define _RISCV_INSN_MACROS_H
+
+// These conflict with Boost headers so can't be included from insn_template.h
+#define P (*p)
+
+#define require(x) do { if (unlikely(!(x))) throw trap_illegal_instruction(insn.bits()); } while (0)
+
+#endif

--- a/riscv/insn_template.cc
+++ b/riscv/insn_template.cc
@@ -1,6 +1,7 @@
 // See LICENSE for license details.
 
 #include "insn_template.h"
+#include "insn_macros.h"
 
 reg_t rv32_NAME(processor_t* p, insn_t insn, reg_t pc)
 {

--- a/riscv/interactive.cc
+++ b/riscv/interactive.cc
@@ -353,19 +353,19 @@ void sim_t::interactive_vreg(const std::string& cmd, const std::vector<std::stri
       uint64_t val;
       switch(elen){
         case 8:
-          val = P.VU.elt<uint64_t>(r, e);
+          val = p->VU.elt<uint64_t>(r, e);
           out << std::dec << "[" << e << "]: 0x" << std::hex << std::setfill ('0') << std::setw(16) << val << "  ";
           break;
         case 4:
-          val = P.VU.elt<uint32_t>(r, e);
+          val = p->VU.elt<uint32_t>(r, e);
           out << std::dec << "[" << e << "]: 0x" << std::hex << std::setfill ('0') << std::setw(8) << (uint32_t)val << "  ";
           break;
         case 2:
-          val = P.VU.elt<uint16_t>(r, e);
+          val = p->VU.elt<uint16_t>(r, e);
           out << std::dec << "[" << e << "]: 0x" << std::hex << std::setfill ('0') << std::setw(8) << (uint16_t)val << "  ";
           break;
         case 1:
-          val = P.VU.elt<uint8_t>(r, e);
+          val = p->VU.elt<uint8_t>(r, e);
           out << std::dec << "[" << e << "]: 0x" << std::hex << std::setfill ('0') << std::setw(8) << (int)(uint8_t)val << "  ";
           break;
       }

